### ISSUE TITLE
[ISSUE] Shipments created while another is processing for the same order

### ISSUE THREAD

**Issue by Sental** [mollie/magento2#656]
**Describe the bug**
If you create a shipment for an order and then while Magento is awaiting a successful response from mollie (or possible while the order is still locked) and a second shipment is sent towards Magento, the order does not fully process correctly. The Order Item does not have it's qty_shipped updated and the order status does not go to complete (in the case of a fully shipped order). I think this could be related to Issue: #655

**Used versions**
- *Magento Version number(eg 2.3.5)*: 2.4.5-p1
- *Open source/Enterprise/B2b*: Open Source
- *Mollie version number (Check configuration)*: 2.23.0

**To Reproduce**
1. Go to the magento admin, create an order and pay for it.
2. Ship one product on the order
3. Immediately ship the rest of the order
4. The shipment is created.
5. The Shipped : 1 line does not appear in the item details and the qty_ordered column is not updated in the database.

**Expected behavior**
The order & order items are all saved as shipped with the status' updated. The shipment in magento is created.

**Actual behavior**
The shipment in magento is created. The qty_shipped for the order lines in the second shipment are not updated and the order status is not updated.

**Screenshots**
Shipment 1:
![shipment-1](https://github.com/mollie/magento2/assets/35261502/11c603dc-d657-45e3-9abd-a2d2608b86a5)
Shipment 2:
![shipment-2](https://github.com/mollie/magento2/assets/35261502/b50d758e-53c4-486c-9861-81fee0c92a4c)
Order Line:
![order-items-list](https://github.com/mollie/magento2/assets/35261502/11fc2295-a932-4aaa-a48b-d473c1a48f9a)

**Additional context**

**Comment by Frank-Magmodules** [comment]
Hello, @Quazz and @Sental! We have taken note of your issues and are currently conducting research on them. We will provide you with further updates shortly. It's possible that we may require additional information later, and if that's the case, we'll inform you accordingly.

**Comment by Frank-Magmodules** [comment]
Hi There! 

We have investigated this again, but we can’t seem to reproduce this. We tried various ways:

*   Manually shipping the order items.
*   Manually shipping the order items in 2 separate tabs at the same time.
*   Shipping the order using the API one after the other (`order/{{order_id}}/ship`)
*   Shipping the order using the API at the same time.

But in all cases, we see the expected behavior. When trying to create two shipments at the same time, we can see in the logs that the second shipment waits on the first shipment to complete, as we expected. 

Could it be that you have some other module that is also invoking the shipment behavior that causes this?

**Comment by Quazz** [comment]
I've noticed something interesting.

We have an extension that logs changes to order data so we can have an easy overview of the order progress.

First we have "Status changed" event

- When a shipment gets created, the state of the order gets updated to complete from processing (good)
- The qty shipped gets updated to the appropriate quantity (good)
- The last transaction ID (in sales_order_payment) gets updated to the shipping transaction id (good)

Then the "Shipment created" event happens

Here's where things get weird.

Most of the time this goes fine and the shipment is simply created and everything is fine.

But sometimes the following happens 

- Changes shipped qty back to the original
- Changes the last trans id (in sales_order_payment) back to the original order transaction id

I don't have that many examples to check at the moment since it only happens sometimes.

Still haven't found a common factor between them, perhaps there isn't one, it does make it impossible to reproduce. 

It's possible it's not caused by Mollie, but it's hard to track down due to its seemingly random nature.

**Comment by Frank-Magmodules** [comment]
Hello @quazz,
As this is super hard for us to reproduce, would it be possible for you to test if that fixes the issue for you? Thanks a lot!

**Comment by Quazz** [comment]
> Hello @Quazz, As this is super hard for us to reproduce, would it be possible for you to test if that fixes the issue for you? Thanks a lot!

@Frank-Magmodules thanks for working on this. I've deployed the new version; now we have to wait and see.

**Comment by Frank-Magmodules** [comment]
Thanks @Quazz , keep me posted.

**Comment by Frank-Magmodules** [comment]
Hi @Quazz , do you have some good news here? ;-)

**Comment by Quazz** [comment]
@Frank-Magmodules Unfortunately, it doesn't appear to have resolved the issue.

**Comment by Frank-Magmodules** [comment]
HI There @Quazz ,

It’s hard to say if that is the issue, but we can try to see if that is the cause. 
Can you install [this](https://gist.github.com/Frank-Magmodules/896e4aac3c4f7c2cbb475f0c570510c8) patch to see if that resolves the issue? 

This calls the `CreateMollieShipment` from within the `LockUnlockOrder` observer. 

_**Please note:** If this fixes the issue there needs to be a better solution in place, but for the purpose to pinpoint the root cause of this issue, this should be fine._

**Comment by Quazz** [comment]
@Frank-Magmodules 

I have applied to patch. We'll see how it goes.

**Comment by Quazz** [comment]
@Frank-Magmodules

Sadly it did not solve the issue (aside from the locking message of course)

**Comment by Quazz** [comment]
The observer gets triggered on `sales_order_shipment_save_before`

At this point the order data has the shipping_item data on 0 (as in not sent).

Eventually the observer reaches https://github.com/mollie/magento2/blob/5636a71c077059e8b4a4d53da133cd5ddc572618/Model/Client/Orders.php#L566 where it saves the order.

https://github.com/magento/magento2/blob/727560d82199f6b938d1906e9d923e2dd40b490a/app/code/Magento/Shipping/Controller/Adminhtml/Order/Shipment/Save.php#L85 controller function plays a role too.

It passes the Order as it is assigned to the shipment by Magento. This means before Mollie modifies it, since the save event for shipment will only get triggered when it reaches save

This explains why there's an orderrepository->save action during an observer event; however this is also potentially why weird problems can arise.

As a result, I will try to put the observer on the _after event instead, this way it should get the updated Magento data without getting in the way of the original save event.

What do you think?

**Comment by Frank-Magmodules** [comment]
Hello @Quazz and @Sental, are you currently experiencing this problem? If so, kindly get in touch with us [directly](https://www.magmodules.eu/help) so that we can exchange environment details and discuss this issue more efficiently.

**Comment by Frank-Magmodules** [comment]
Good morning, @Quazz and @Sental,

I haven't received any updates on this issue. Would it be possible for us to explore this together, or should we consider closing the matter for the time being?

**Comment by Frank-Magmodules** [comment]
Hi There @Quazz  and @Sental ,

Because of inactivity, I am closing this matter for the time being. If you have any further questions or concerns regarding this issue, please don't hesitate to reopen this ticket or reach out to us [directly](https://www.magmodules.eu/support-form.html).

**Comment by mokadev** [comment]
Hi @Frank-Magmodules 

I have the same problem and this patch https://gist.github.com/Frank-Magmodules/896e4aac3c4f7c2cbb475f0c570510c8  didn't work. (I installed the last version of the module) 

The problem happened when the invoice is created by Mollie, I have an observer on "sales_order_invoice_register" event which create a shipment and a shipping label. 

It seems that the lock placed by mollie prevent throw an exception "Mollie.ERROR: error: Rolled back transaction has not been completed correctly":

`[2024-04-30T20:32:36.124848+00:00] Mollie.INFO: info: Locking: mollie.order.220180 [] []
[2024-04-30T20:32:36.428620+00:00] Mollie.INFO: webhook: {"resource":"payment","id":"tr_yneoRi5E6x","mode":"test","amount":{"value":"87.40","currency":"EUR"},"settlementAmount":{"value":"87.40","currency":"EUR"},"amountRefunded":{"value":"0.00","currency":"EUR"},"amountRemaining":{"value":"87.40","currency":"EUR"},"amountChargedBack":null,"description":"100287787","method":"creditcard","status":"paid","createdAt":"2024-04-30T20:28:58+00:00","paidAt":"2024-04-30T20:29:03+00:00",[...],"countryCode":"FR"} [] []
[2024-04-30T20:32:36.490369+00:00] Mollie.INFO: info: Lock for "mollie.order.220180" is already active, attempt 0 (sleep for: 0.5) [] []
[2024-04-30T20:32:36.990884+00:00] Mollie.INFO: info: Lock for "mollie.order.220180" is already active, attempt 1 (sleep for: 1.0) [] []
[2024-04-30T20:32:37.991415+00:00] Mollie.INFO: info: Lock for "mollie.order.220180" is already active, attempt 2 (sleep for: 1.5) [] []
[2024-04-30T20:32:39.491950+00:00] Mollie.INFO: info: Lock for "mollie.order.220180" is already active, attempt 3 (sleep for: 2.0) [] []
[2024-04-30T20:32:41.492487+00:00] Mollie.INFO: info: Lock for "mollie.order.220180" is already active, attempt 4 (sleep for: 2.5) [] []
[2024-04-30T20:32:43.992980+00:00] Mollie.INFO: info: Lock for "mollie.order.220180" is already active, attempt 5 (sleep for: 3.0) [] []
[2024-04-30T20:32:46.998544+00:00] Mollie.INFO: info: Unlocking: mollie.order.220180 [] []
[2024-04-30T20:32:46.998743+00:00] Mollie.INFO: info: Key "mollie.order.220180" unlocked [] []
[2024-04-30T20:32:46.998797+00:00] Mollie.ERROR: error: Rolled back transaction has not been completed correctly. [] []`

It seems that we can't create a shipment just after the invoice is created.

**Comment by Frank-Magmodules** [comment]
Hello @mokadev (and others in this issue),

The error indicates that there was an exception somewhere down the line. It’s very unfortunate that it didn’t get logged in these logs though. Would it be possible to check if it’s the exception.log of your webshop or any other log file? That might give some information about why this error is occurring. 

If there are no errors, could you describe or share what kind of mutations you are doing on the order and/or shipping objects? This might help us reproduce this issue.

**Comment by mokadev** [comment]
Hi @Frank-Magmodules no exception found . 

For now to work around the bug i removed my shipping label generation after invoice creation. So the problem is still there (weird that the module lock the order even if Mollie is not enabled) but it doesn't block me anymore.

Thanks

**Comment by Frank-Magmodules** [comment]
Great @mokadev , thanks for your response. I'm closing this for now. If you have a specific issue or request, please open a new issue with a more detailed description so we can address it more effectively.
